apps/extension/src/core/managers/config.ts

```diff
@@ -1,98 +1,256 @@
     this.defaultConfig = new Storage({
-      area: "local"
+      area: "sync"
     })
     this.defaultConfig.setNamespace(`configs-default-`)
+
+    // For each ModelID, store the id of the config that handles it
+    this.modelHandlers = new Storage({
+      area: "sync"
+    })
+    this.modelHandlers.setNamespace(`configs-model-handlers-`)
+  }
+
+  init(auth: AuthType, modelId?: ModelID): Config {
+    const id = uuidv4()
+    const caller = this.getCallerForAuth(auth, modelId)
+    const label = this.getLabelForAuth(auth, modelId)
+    switch (auth) {
+      case AuthType.External:
+        return {
+          id,
+          auth,
+          label,
+          models: [ModelID.GPT3, ModelID.GPT4],
+          baseUrl: caller.config.defaultBaseUrl
+        }
+      case AuthType.APIKey:
+        return {
+          id,
+          auth,
+          models: modelId ? [modelId] : [],
+          baseUrl: caller.config.defaultBaseUrl,
+          label
+        }
+    }
   }
 
-  init(id: ModelID): Config {
-    return {
-      id,
-      baseUrl: modelCallers[id].config.defaultBaseUrl,
-      label: defaultLabel[id]
+  async save(config: Config): Promise<boolean> {
+    const isNew = await super.save(config)
+
+    if (isNew) {
+      // Index by auth type
+      await this.indexBy(config, config.auth, authIndexName)
+    }
+
+    for (const modelId of config.models) {
+      await this.modelHandlers.set(modelId, config.id)
     }
+
+    return isNew
   }
 
-  async get(id: ModelID): Promise<Config | undefined> {
-    const obj = await super.get(id)
-    const defaults = this.init(id)
-    return {
-      ...defaults,
-      ...obj
+  async forModel(modelId: ModelID): Promise<Config> {
+    const configId = await this.modelHandlers.get(modelId)
+    if (configId) {
+      const config = await this.get(configId)
+      if (config) {
+        const defaults = this.init(config.auth, modelId)
+        return {
+          ...defaults,
+          ...config
+        }
+      }
+      await this.modelHandlers.remove(modelId)
     }
+    // TODO include Token auth possibilities?
+    return this.init(AuthType.APIKey, modelId)
   }
 
-  isIncomplete(config: Config): boolean {
-    return (
-      !config.baseUrl ||
-      (![ModelID.Local, ModelID.Together].includes(config.id) && !config.apiKey)
-    )
+  isCredentialed(config: Config): boolean {
+    if (!config.baseUrl) {
+      return false
+    }
+    switch (config.auth) {
+      case AuthType.External:
+        return !!config.authMetadata
+      case AuthType.APIKey:
+        return config.models.length ? !!config.apiKey : true
+    }
   }
 
-  async setDefault(id: ModelID) {
-    const previous = (await this.defaultConfig.get("id")) as ModelID | undefined
-    await this.defaultConfig.set("id", id)
-    if (previous !== id) {
+  async setDefault(config: Config) {
+    await this.save(config)
+    const previous = (await this.defaultConfig.get("id")) as string | undefined
```

Comment:
Maybe we should move this into a `getId` method inside that BaseManager class, since there's already a `getIds` method.

Reply:
wouldn't work here i think because `defaultConfig` is a new Storage instance that's specially managed by this manager

Reply:
oh whoops lol - the code inspector pointed me to the BaseManager class somehow when I was looking up the interface of defaultConfig. 

If that's the case, I don't think you need to do the type cast since the default return type of Storage.get is `string`. `T | undefined` is the same as `T` since undefined is implicit:

<img width="501" alt="Screenshot 2023-04-29 at 4 59 26 PM" src="https://user-images.githubusercontent.com/6723574/235324013-2d4bcffa-72ad-4058-8f12-62d9969f0708.png">

Reply:
ah right, weird that i had that